### PR TITLE
Added ability to send the client job heartbeat calls to server.

### PR DESCRIPTION
**Pull request by yhwen**
Fixes # .

### Description

Added the ability to periodically send the heartbeat calls to the server from the client job process.

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Quick tests passed locally by running `./runtest.sh`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated.

### CODE REVIEW

File: nvflare/private/fed/client/client_runner.py [hunk 1]
```diff
@@ -404,6 +407,17 @@ def _try_run(self):
 
                 time.sleep(task_fetch_interval)
 
+    def send_job_heartbeat(self, interval=30.0):
+        wait_times = int(interval / 2)
+        while not self.asked_to_stop:
+            with self.engine.new_context() as fl_ctx:
+                self.engine.send_job_heartbeat(fl_ctx)
+
+                for i in range(wait_times):
```
Comment (nvidianz) [inline nvflare/private/fed/client/client_runner.py:416]: An event might work better here. An example can be found here: https://github.com/NVIDIA/NVFlare/blob/ab0c7296cbe57568ec0f7c1f07c321eeb44f9787/nvflare/fuel/f3/sfm/heartbeat_monitor.py#L47-L54

**Review by nvidianz — APPROVED** [review]

File: nvflare/private/fed/client/client_runner.py [hunk 2]
```diff
@@ -392,6 +392,9 @@ def _check_stop_conditions(self, fl_ctx: FLContext) -> bool:
         return False
 
     def _try_run(self):
+        heartbeat_thread = threading.Thread(target=self.send_job_heartbeat, args=[])
```
Comment (yanchengnv) [inline nvflare/private/fed/client/client_runner.py:395]: Should set it to daemon

File: nvflare/private/fed/client/client_run_manager.py [hunk 3]
```diff
@@ -115,6 +115,9 @@ def __init__(
 
         self.logger = logging.getLogger(self.__class__.__name__)
 
+    def send_job_heartbeat(self, fl_ctx: FLContext):
```
Comment (yanchengnv) [inline nvflare/private/fed/client/client_run_manager.py:118]: Why do we need this? No one will call this method.

File: nvflare/private/fed/client/client_engine_executor_spec.py [hunk 4]
```diff
@@ -45,6 +45,10 @@ def __init__(self, name: str, task_id: str, data: Shareable):
 class ClientEngineExecutorSpec(ClientEngineSpec, EngineSpec, ABC):
     """The ClientEngineExecutorSpec defines the ClientEngine APIs running in the child process."""
 
+    @abstractmethod
+    def send_job_heartbeat(self, fl_ctx: FLContext):
```
Comment (yanchengnv) [inline nvflare/private/fed/client/client_engine_executor_spec.py:49]: Why do we need this in spec?

File: nvflare/private/fed/client/client_runner.py [hunk 5]
```diff
@@ -404,6 +407,17 @@ def _try_run(self):
 
                 time.sleep(task_fetch_interval)
 
+    def send_job_heartbeat(self, interval=30.0):
+        wait_times = int(interval / 2)
+        while not self.asked_to_stop:
+            with self.engine.new_context() as fl_ctx:
+                self.engine.send_job_heartbeat(fl_ctx)
```
Comment (yanchengnv) [inline nvflare/private/fed/client/client_runner.py:414]: Should keep it simple - just send an aux message,

File: nvflare/private/fed/client/communicator.py [hunk 6]
```diff
@@ -367,6 +367,31 @@ def send_heartbeat(self, servers, task_name, token, ssid, client_name, engine: C
                 self.logger.info(f"Failed to send heartbeat. Will try again. Exception: {secure_format_exception(e)}")
                 time.sleep(5)
 
+    def send_job_heartbeat(self, token, ssid, client_name, fl_ctx: FLContext):
```
Comment (yanchengnv) [inline nvflare/private/fed/client/communicator.py:370]: No need for this if we simply send aux msg from client_runner.

File: nvflare/private/fed/client/fed_client_base.py [hunk 7]
```diff
@@ -324,6 +324,12 @@ def send_heartbeat(self, project_name, interval):
         except FLCommunicationError:
             self.communicator.heartbeat_done = True
 
+    def send_job_heartbeat(self, project_name, fl_ctx: FLContext):
```
Comment (yanchengnv) [inline nvflare/private/fed/client/fed_client_base.py:327]: No need for this.

File: nvflare/private/fed/client/fed_client_base.py [hunk 8]
```diff
@@ -345,6 +351,17 @@ def heartbeat(self, interval):
             if pool:
                 pool.terminate()
 
+    def run_job_heartbeat(self, fl_ctx: FLContext):
```
Comment (yanchengnv) [inline nvflare/private/fed/client/fed_client_base.py:354]: No need.

File: nvflare/private/fed/server/fed_server.py [hunk 9]
```diff
@@ -327,6 +327,11 @@ def _register_cellnet_cbs(self):
             topic=CellChannelTopic.HEART_BEAT,
             cb=self.client_heartbeat,
         )
+        self.cell.register_request_cb(
```
Comment (yanchengnv) [inline nvflare/private/fed/server/fed_server.py:330]: No need here.
Should just register a CB in server_runner.

**Review by yanchengnv — CHANGES_REQUESTED** [review]
This PR is more complicated than necessary. I think we just simply need an aux message between client_runner and server_runner. No need to touch any of the engine stuff.

Please keep it simple.

**Comment by yhwen** [comment]
> This PR is more complicated than necessary. I think we just simply need an aux message between client_runner and server_runner. No need to touch any of the engine stuff.
> 
> Please keep it simple.

Changed to use the aux_message for sending the client job heartbeat calls.

File: nvflare/private/fed/server/fed_server.py (hunk 9 above)
Comment (yhwen) [reply nvflare/private/fed/server/fed_server.py]: changed to use aux_message sending.

File: nvflare/private/fed/client/fed_client_base.py (hunk 8 above)
Comment (yhwen) [reply nvflare/private/fed/client/fed_client_base.py]: Removed.

File: nvflare/private/fed/client/fed_client_base.py (hunk 7 above)
Comment (yhwen) [reply nvflare/private/fed/client/fed_client_base.py]: Removed.

File: nvflare/private/fed/client/communicator.py (hunk 6 above)
Comment (yhwen) [reply nvflare/private/fed/client/communicator.py]: Removed. Changed to use aux_message sending.

File: nvflare/private/fed/client/client_engine_executor_spec.py (hunk 4 above)
Comment (yhwen) [reply nvflare/private/fed/client/client_engine_executor_spec.py]: Removed.

File: nvflare/private/fed/client/client_run_manager.py (hunk 3 above)
Comment (yhwen) [reply nvflare/private/fed/client/client_run_manager.py]: Removed.

File: nvflare/private/fed/client/client_runner.py (hunk 2 above)
Comment (yhwen) [reply nvflare/private/fed/client/client_runner.py]: set it to daemon.

File: nvflare/private/fed/client/client_runner.py (hunk 1 above)
Comment (yhwen) [reply nvflare/private/fed/client/client_runner.py:425]: Keep the time interval checking for now.

**Review by yanchengnv — APPROVED** [review]
LGTM

**Pull request merged**